### PR TITLE
Fix slicing

### LINKED ISSUE #34: Slicing with a single index raises error
Example:

```Py
from scippnexus import data
filename = data.get_path('PG3_4844_event.nxs')
import scippnexus as snx
f = snx.File(filename)

data = f['entry/bank103']

data['x_pixel_offset', 0]
```
raises
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 data['x_pixel_offset', 0]

File ~/code/nxus/jupyter/scippnexus/nxobject.py:315, in NXobject.__getitem__(self, name)
    312 def __getitem__(
    313         self,
    314         name: NXobjectIndex) -> Union['NXobject', Field, sc.DataArray, sc.Dataset]:
--> 315     return self._get_child(name, use_field_dims=True)

File ~/code/nxus/jupyter/scippnexus/nxobject.py:307, in NXobject._get_child(self, name, use_field_dims)
    305     else:
    306         return _make(item)
--> 307 da = self._getitem(name)
    308 if (t := self.depends_on) is not None:
    309     da.coords['depends_on'] = t if isinstance(t, sc.Variable) else sc.scalar(t)

File ~/code/nxus/jupyter/scippnexus/nxdata.py:149, in NXdata._getitem(self, select)
    148 def _getitem(self, select: ScippIndex) -> sc.DataArray:
--> 149     signal = self._signal[select]
    150     if self._errors_name in self:
    151         stddevs = self[self._errors_name][select]

File ~/code/nxus/jupyter/scippnexus/nxobject.py:185, in Field.__getitem__(self, select)
    183 shape = list(self.shape)
    184 for i, ind in enumerate(index):
--> 185     shape[i] = len(range(*ind.indices(shape[i])))
    187 variable = sc.empty(dims=self.dims,
    188                     shape=shape,
    189                     dtype=self.dtype,
    190                     unit=self.unit)
    192 # If the variable is empty, return early

AttributeError: 'int' object has no attribute 'indices'
```

### LINKED ISSUE #35: Slicing with bin edge coord does not preserve coord as edges

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #34.
Fixes #35.

### CODE REVIEW

File: src/scippnexus/_common.py [hunk 1]
```diff
@@ -94,18 +94,26 @@ def to_plain_index(dims: List[str], select: ScippIndex) -> Union[int, slice, tup
...
 
     This removes any selections that apply to the parent but not the child.
     """
-    if set(dims) == set(child_dims):
-        return select
     select = _to_canonical_select(dims, select)
-    for d in dims:
-        if d not in child_dims and d in select:
-            del select[d]
+    if set(dims) != set(child_dims):
```
Comment (jl-wynen) [inline src/scippnexus/_common.py:108]: This `if` can be removed.

File: src/scippnexus/_common.py [hunk 2]
```diff
@@ -94,18 +94,26 @@ def to_plain_index(dims: List[str], select: ScippIndex) -> Union[int, slice, tup
...
+        for d in dims:
+            if d not in child_dims and d in select:
+                del select[d]
+    for dim in select:
+        if dim == bin_edge_dim:
+            index = select[dim]
+            if isinstance(index, int):
+                select[dim] = slice(index, index + 2)
+            else:
+                select[dim] = slice(index.start, index.stop + 1)
```
Comment (jl-wynen) [inline src/scippnexus/_common.py:118]: Why is it `+1` in the second case but `+2` in the first?

File: src/scippnexus/nxdata.py [hunk 3]
```diff
@@ -171,7 +196,11 @@ def _getitem(self, select: ScippIndex) -> sc.DataArray:
                                  f"{name}_error. The latter will be ignored.")
                         stddevs = self[f'{name}{suffix}'][sel]
                         coord.variances = sc.pow(stddevs, sc.scalar(2)).values
-                da.coords[name] = coord
+                if self._coord_to_attr(da, name, field):
+                    # Like scipp, slicing turns dim
```
Comment (jl-wynen) [inline src/scippnexus/nxdata.py:200]: Missing part of the sentence?

File: tests/nxdata_test.py [hunk 4]
```diff
@@ -378,3 +378,22 @@ def test_fields_with_datetime_attribute_are_loaded_as_datetime(nxroot):
...
+    data.create_field('data', da.data)
+    data.attrs['signal'] = 'data'
+    data.attrs['axes'] = ['xx']
+    data.attrs['xx_indices'] = [0]
+    data.attrs['xx2_indices'] = [0]
+    assert sc.identical(data[...], da)
+    assert sc.identical(data['xx', 0], da['xx', 0])
+    assert sc.identical(data['xx', 1], da['xx', 1])
+    assert sc.identical(data['xx', 0:1], da['xx', 0:1])
+    assert sc.identical(data['xx', 1:3], da['xx', 1:3])
```
Comment (jl-wynen) [inline tests/nxdata_test.py:399]: Add a test with an empty slice.

File: tests/nxdetector_test.py [hunk 5]
```diff
@@ -84,6 +84,38 @@ def test_loads_data_with_coords(nxroot):
...
+    detector.create_field('detector_numbers', da.coords['detector_numbers'])
+    detector.create_field('xx', da.coords['xx'])
+    detector.create_field('xx2', da.coords['xx2'])
+    detector.create_field('yy', da.coords['yy'])
+    detector.create_field('2d_edges', da.coords['2d_edges'])
+    detector.create_field('data', da.data)
+    detector.attrs['axes'] = ['xx', 'yy']
+    detector.attrs['2d_edges_indices'] = [1, 0]
+    assert sc.identical(detector[...], da)
+    print(detector['xx', 0], da['xx', 0])
```
Comment (jl-wynen) [inline tests/nxdetector_test.py:110]:
```suggestion
```

File: tests/nxdetector_test.py [hunk 6]
```diff
@@ -84,6 +84,38 @@ def test_loads_data_with_coords(nxroot):
...
+    assert sc.identical(detector['xx', 0], da['xx', 0])
+    assert sc.identical(detector['xx', 1], da['xx', 1])
+    assert sc.identical(detector['xx', 0:1], da['xx', 0:1])
+    assert sc.identical(detector['yy', 0], da['yy', 0])
+    assert sc.identical(detector['yy', 1], da['yy', 1])
+    assert sc.identical(detector['yy', 0:1], da['yy', 0:1])
```
Comment (jl-wynen) [inline tests/nxdetector_test.py:116]: Again, add a test with an empty slice.

File: src/scippnexus/_common.py (hunk 2 above)
Comment (SimonHeybrock) [reply src/scippnexus/_common.py:118]: Consider extracting a "thickness 1" slice, so we need to edges. Two cases:

- `obj[i]` => edges `i:i+2`
- `obj[i:i+1] => edges `i:i+2`.

Since `index.stop == i+1`, use use `stop + 1`

Comment (jl-wynen) [reply src/scippnexus/_common.py:118]: ok

File: tests/nxdata_test.py (hunk 4 above)
Comment (SimonHeybrock) [reply tests/nxdata_test.py:399]: Good catch, it is indeed broken.

**Review by jl-wynen — APPROVED** [review]
